File: go/gf_apps/gf_images_lib/gf_images_core/gf_images_share.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2023 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_images_core

import (
	"fmt"
	"context"
	"github.com/gloflow/gloflow/go/gf_core"
	"github.com/gloflow/gloflow/go/gf_events"
	"github.com/gloflow/gloflow/go/gf_extern_services/gf_aws"
	"github.com/gloflow/gloflow/go/gf_identity/gf_identity_core"
)

//---------------------------------------------------

type GFshareInput struct {
	ImageID         GFimageID
	EmailAddressStr string
	EmailSubjectStr string
	EmailBodyStr    string
}

//---------------------------------------------------

func SharePipeline(pInput *GFshareInput,
	pUserID      gf_core.GF_ID,
	pServiceInfo *GFserviceInfo,
	pCtx         context.Context,
	pRuntimeSys  *gf_core.RuntimeSys) *gf_core.GFerror {

	//------------------------
	// DB
	
	/*
	// SENDER_EMAIL_ADDRESS
	senderAddressStr, gfErr := gf_identity_core.DBsqlGetUserEmailByID(pUserID,
		pCtx,
		pRuntimeSys)
	if gfErr != nil {
		return gfErr
	}
	*/

	// SENDER_USER_NAME
	userNameStr, gfErr := gf_identity_core.DBsqlGetUserNameByID(pUserID,
		pCtx,
		pRuntimeSys)
	if gfErr != nil {
		return gfErr
	}
	
	//------------------------
	// IMAGE

	image, gfErr := DBgetImage(pInput.ImageID, pCtx, pRuntimeSys)
	if gfErr != nil {
		return gfErr
	}
	
	/*
	image, gfErr := DBmongoGetImage(pInput.ImageID, pCtx, pRuntimeSys)
	if gfErr != nil {
		return gfErr
	}
	*/

	//------------------------

	imageURLstr := fmt.Sprintf("https://%s%s", pServiceInfo.DomainBaseStr, image.ThumbnailMediumURLstr)

	//------------------------

	senderAddressStr := pServiceInfo.EmailSharingSenderAddressStr

	msgBodyHTMLstr := fmt.Sprintf(`
		<div>
			<div style="margin-top: 20px;">

				<a href="https://%s" style="color:gray;">gloflow</a> user <span style="
				background-color: #e1e1e1;
				padding: 6px;
				padding-left: 9px;
				padding-right: 8px;
				font-weight: bold;">%s</span> shared this image with you :) 
			</div>
			<div id='user_body'>
				%s
			</div>
			<div id='image' style="margin-top: 40px;">
				<img src='%s' alt='image' style='width:50%%;'></img>
			</div>
		</div>`,
		pServiceInfo.DomainBaseStr,
		userNameStr,
		pInput.EmailBodyStr,
		imageURLstr)

	//------------------------
	// AWS
	gfErr = gf_aws.SESsendMessage(pInput.EmailAddressStr,
		senderAddressStr,
		pInput.EmailSubjectStr,
		msgBodyHTMLstr,
		"", // msgBodyTextStr,
		pRuntimeSys)
	
	if gfErr != nil {
		return gfErr
	}

	//------------------------
	// PLUGIN - email sending

	emailGFtypeStr := "image_sharing"
	
	gfErr = pRuntimeSys.ExternalPlugins.EmailSendingCallback(pInput.EmailAddressStr,
		senderAddressStr,
		pInput.EmailSubjectStr,
		emailGFtypeStr,
		pRuntimeSys)
	if gfErr != nil {
		return gfErr
	}
	
	//------------------------
	// EVENT
	if pServiceInfo.EnableEventsAppBool {
		eventMetaMap := map[string]interface{}{
			"image_id": pInput.ImageID,
			"user_id":  pUserID,
			"email":    pInput.EmailAddressStr,
		}
		gf_events.EmitApp(GF_EVENT_APP__IMAGE_SHARE,
			eventMetaMap,
			pRuntimeSys.AppNameStr,
			pUserID,
			pCtx,
			pRuntimeSys)
	}

	//------------------------
	

	return nil
}

File: web/src/gf_apps/gf_images/ts/gf_images_core/gf_images_share.ts
/*
GloFlow application and media management/publishing platform
Copyright (C) 2024 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

import * as gf_images_http from "./gf_images_http";

//---------------------------------------------------
export function init(p_image_id_str,
	p_image_container_element,
	p_gf_host_str,
	p_log_fun) {



	const http_api_map = {
		"gf_images": {
			"share_image": async (p_image_id_str,
				p_email_address_str,
                p_email_subject_str,
                p_email_body_str,
				p_log_fun)=>{
				const p = new Promise(async function(p_resolve_fun, p_reject_fun) {

					await gf_images_http.share(p_image_id_str,
                        p_email_address_str,
                        p_email_subject_str,
                        p_email_body_str,
						p_gf_host_str,
						p_log_fun);

					p_resolve_fun({
						
					});
				});
				return p;
			}
		}
	};

    // UI
    const control_element = init_ui(p_image_id_str,
		p_image_container_element,
		http_api_map,
        p_log_fun);


    $(p_image_container_element).append(control_element);
}

//---------------------------------------------------

function init_ui(p_image_id_str,
	p_image_container_element,
	p_http_api_map,
    p_log_fun) {





    const control_element = $(`
        <div class="gf_images_share">
            <div class="email_btn">
				S
            </div>
        </div>`);


	var opened_bool = false;
	var sharing_dialog_element;

	$(control_element).click(()=>{

		if(opened_bool){
			$(sharing_dialog_element).remove();
			opened_bool = false;
		}
		else {
			sharing_dialog_element = email_share_dialog();

			$(control_element).append(sharing_dialog_element);
			opened_bool = true;
		}

	});

	//------------------------
	// IMPORTANT!! - onMouseEnter/onMouseLeave fire when the target element is entered/left, 
	//               but unline mouseon/mouseout it will not fire if its children are entered/left.
	
	$(p_image_container_element).on('mouseenter', (p_event)=>{
		$(p_image_container_element).append(control_element);
	});

	$(p_image_container_element).on('mouseleave', (p_event)=>{

        // IMPORTANT!! - detaching in order to keep event handlers
		$(control_element).detach();
	});

	//---------------------------------------------------
	function email_share_dialog() {

		const dialog = $(`
		<div class="gf_email_dialog">
		
			<div class="email_address">
				<input type="text" placeholder="email address">
			</div>
			<div class="email_subject">
				<input type="text" placeholder="email subject">
			</div>
			<div class="email_body">
				<textarea placeholder="email body"></textarea>
			</div>
			<div class="share_btn">
				share
			</div>
		</div>`);


		$(dialog).find(".share_btn").click(async ()=>{

			const email_address_str = $(dialog).find(".email_address input").val();
			const email_subject_str = $(dialog).find(".email_subject input").val();
			const email_body_str = $(dialog).find(".email_body textarea").val();

			// HTTP
			await p_http_api_map.gf_images.share_image(p_image_id_str,
				email_address_str,
				email_subject_str,
				email_body_str,
				p_log_fun);
			
			// mark the button as complete
			$(dialog).find(".share_btn").css("background-color", "green");

			setTimeout(()=>{
					$(dialog).remove();
				}, 1000);
		});


		return dialog;
	}

	//---------------------------------------------------

    return control_element;
}